File: java/com/epicdragonworld/gameserver/model/actor/Player.java
package com.epicdragonworld.gameserver.model.actor;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.logging.Logger;

import com.epicdragonworld.gameserver.managers.DatabaseManager;
import com.epicdragonworld.gameserver.network.GameClient;
import com.epicdragonworld.gameserver.network.SendablePacket;

/**
 * @author Pantelis Andrianakis
 */
public class Player extends Creature
{
	private static final Logger LOGGER = Logger.getLogger(Player.class.getName());
	private static final String RESTORE_CHARACTER = "SELECT * FROM characters WHERE name=?";
	private static final String STORE_CHARACTER = "UPDATE characters SET name=?, class_id=? WHERE account=? AND name=?";
	
	private final GameClient _client;
	private final String _name;
	private int _classId = 0;
	
	public Player(GameClient client, String name)
	{
		_client = client;
		_name = name;
		
		// Load information from database.
		try (Connection con = DatabaseManager.getConnection();
			PreparedStatement ps = con.prepareStatement(RESTORE_CHARACTER))
		{
			ps.setString(1, name);
			try (ResultSet rset = ps.executeQuery())
			{
				while (rset.next())
				{
					_classId = rset.getInt("class_id");
					getLocation().setX(rset.getFloat("x"));
					getLocation().setY(rset.getFloat("y"));
					getLocation().setZ(rset.getFloat("z"));
					// TODO: Restore player stats (STA/STR/DEX/INT).
					// TODO: Restore player level.
					// TODO: Restore player Current HP.
					// TODO: Restore player Current MP.
				}
			}
		}
		catch (Exception e)
		{
			LOGGER.warning(e.getMessage());
		}
	}
	
	public void storeMe()
	{
		try (Connection con = DatabaseManager.getConnection();
			PreparedStatement ps = con.prepareStatement(STORE_CHARACTER))
		{
			ps.setString(1, _name);
			ps.setInt(2, _classId);
			// TODO: Save location.
			ps.setString(3, _client.getAccountName());
			ps.setString(4, _name);
			ps.execute();
		}
		catch (Exception e)
		{
			LOGGER.warning(e.getMessage());
		}
	}
	
	public GameClient getClient()
	{
		return _client;
	}
	
	public String getName()
	{
		return _name;
	}
	
	public int getClassId()
	{
		return _classId;
	}
	
	@Override
	public boolean isPlayer()
	{
		return true;
	}
	
	@Override
	public Player asPlayer()
	{
		return this;
	}
	
	public void channelSend(SendablePacket packet)
	{
		_client.channelSend(packet);
	}
}

File: EpicDragonWorldServer/Actor/Player.cs
using MySql.Data.MySqlClient;
using System;

/**
 * Author: Pantelis Andrianakis
 * Date: November 7th 2018
 */
public class Player : Creature
{
    private static readonly string RESTORE_CHARACTER = "SELECT * FROM characters WHERE name=@name";
    private static readonly string STORE_CHARACTER = "UPDATE characters SET name=@name, race=@race, x=@x, y=@y, z=@z, heading=@heading, experience=@experience, hp=@hp, mp=@mp WHERE account=@account AND name=@name";

    private readonly GameClient _client;
    private readonly string _name;
    private readonly byte _raceId;
    private readonly float _height;
    private readonly float _belly;
    private readonly int _hairType;
    private readonly int _hairColor;
    private readonly int _skinColor;
    private readonly int _eyeColor;
    private readonly long _experience;
    private readonly byte _accessLevel;
    private readonly Inventory _inventory;

    public Player(GameClient client, string name)
    {
        _client = client;
        _name = name;

        // Load information from database.
        try
        {
            MySqlConnection con = DatabaseManager.GetConnection();
            MySqlCommand cmd = new MySqlCommand(RESTORE_CHARACTER, con);
            cmd.Parameters.AddWithValue("name", name);
            MySqlDataReader reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                _raceId = (byte)reader.GetInt16("race"); // TODO: Remove cast?
                _height = (float)reader.GetDouble("height"); // Fixes known MySQL float issue.
                _belly = (float)reader.GetDouble("belly"); // Fixes known MySQL float issue.
                _hairType = (byte)reader.GetInt16("hair_type"); // TODO: Remove cast?
                _hairColor = reader.GetInt32("hair_color");
                _skinColor = reader.GetInt32("skin_color");
                _eyeColor = reader.GetInt32("eye_color");

                float locX = (float)reader.GetDouble("x"); // Fixes known MySQL float issue.
                float locY = (float)reader.GetDouble("y"); // Fixes known MySQL float issue.
                float locZ = (float)reader.GetDouble("z"); // Fixes known MySQL float issue.

                // Check if player is outside of world bounds.
                if (locX < Config.WORLD_MINIMUM_X || locX > Config.WORLD_MAXIMUM_X || locY < Config.WORLD_MINIMUM_Y || locY > Config.WORLD_MAXIMUM_Y || locZ < Config.WORLD_MINIMUM_Z || locZ > Config.WORLD_MAXIMUM_Z)
                {
                    // Move to initial area.
                    SetLocation(Config.STARTING_LOCATION);
                }
                else
                {
                    SetLocation(new LocationHolder(locX, locY, locZ, (float)reader.GetDouble("heading"))); // Fixes known MySQL float issue.
                }

                _experience = reader.GetInt64("experience");
                SetCurrentHp(reader.GetInt64("hp"));
                SetCurrentMp(reader.GetInt64("mp"));
                _accessLevel = (byte)reader.GetInt16("access_level"); // TODO: Remove cast?
            }
            con.Close();
        }
        catch (Exception e)
        {
            LogManager.Log(e.ToString());
        }

        // Initialize inventory.
        _inventory = new Inventory(name);
    }

    public void StoreMe()
    {
        try
        {
            MySqlConnection con = DatabaseManager.GetConnection();
            MySqlCommand cmd = new MySqlCommand(STORE_CHARACTER, con);
            cmd.Parameters.AddWithValue("name", _name);
            cmd.Parameters.AddWithValue("race", _raceId);
            cmd.Parameters.AddWithValue("x", GetLocation().GetX());
            cmd.Parameters.AddWithValue("y", GetLocation().GetY());
            cmd.Parameters.AddWithValue("z", GetLocation().GetZ());
            cmd.Parameters.AddWithValue("heading", GetLocation().GetHeading());
            cmd.Parameters.AddWithValue("experience", _experience);
            cmd.Parameters.AddWithValue("hp", GetCurrentHp());
            cmd.Parameters.AddWithValue("mp", GetCurrentMp());
            cmd.Parameters.AddWithValue("account", _client.GetAccountName());
            cmd.ExecuteNonQuery();
            con.Close();
        }
        catch (Exception e)
        {
            LogManager.Log(e.ToString());
        }

        // Save inventory.
        _inventory.Store(_name);
    }

    public GameClient GetClient()
    {
        return _client;
    }

    public string GetName()
    {
        return _name;
    }

    public int GetRaceId()
    {
        return _raceId;
    }

    public float GetHeight()
    {
        return _height;
    }

    public float GetBelly()
    {
        return _belly;
    }

    public int GetHairType()
    {
        return _hairType;
    }

    public int GetHairColor()
    {
        return _hairColor;
    }

    public int GetSkinColor()
    {
        return _skinColor;
    }

    public int GetEyeColor()
    {
        return _eyeColor;
    }

    public long GetExperience()
    {
        return _experience;
    }

    public byte GetAccessLevel()
    {
        return _accessLevel;
    }

    public Inventory GetInventory()
    {
        return _inventory;
    }

    public void ChannelSend(SendablePacket packet)
    {
        _client.ChannelSend(packet);
    }

    public override bool IsPlayer()
    {
        return true;
    }

    public override Player AsPlayer()
    {
        return this;
    }

    public override string ToString()
    {
        return "Player [" + _name + "]";
    }
}
